File: pantry-tracker-backend/database_operations/modifyIngredientEntry.py
from extensions import db
from sqlalchemy import text
from flask import Blueprint, request, jsonify

modifyIngredientEntry_bp = Blueprint('modifyIngredientEntry', __name__)

@modifyIngredientEntry_bp.route('/modify-ingredient-entry', methods=['POST'])

def modify_ingredient_entry():
    data = request.get_json()
    ingredientName = data.get('ingredientName')
    quantity = data.get('quantity')
    units = data.get('units')

    try:
        isIngredient = db.session.execute(text('SELECT aliased_ingredient_name FROM my_pantry WHERE aliased_ingredient_name ILIKE :ingredientName'), {'ingredientName': f'%{ingredientName}%'}).fetchone()
        unitID = db.session.execute(text('SELECT id FROM units WHERE unit_name = :units'), {'units': units}).scalar()
        if (isIngredient):
            db.session.execute(text('UPDATE my_pantry SET quantity = :quantity, unit_id = :unitID WHERE aliased_ingredient_name = :ingredientName'), {'quantity': quantity, 'unitID': unitID, 'ingredientName': ingredientName})
            db.session.commit()
        else:
            return jsonify({'error': 'Ingredient not found'}), 404
        
        return jsonify({'message': 'Ingredient entry modified successfully'}), 200
        
    except Exception as e:
        return jsonify({'error': str(e)}), 500

File: pantry-tracker-frontend/src/supabase_db_operations/modifyIngredientEntry.js
import supabaseClient from '../utils/SupabaseClient';
import getUserID from '../utils/getUserID';
// import API_BASE from '../utils/Api';

const modifyIngredientEntry = async ({ ingredientName, quantity, units }) => {
    const userID = await getUserID();
    try {
        // check if entry exists in db
        const { data: ingredientExists, error: ingredientExistsError } = await supabaseClient
            .from('user_ingredients')
            .select('aliased_ingredient_name, p_unit_id')
            .eq('aliased_ingredient_name', ingredientName)
            .eq('user_id', userID);


        if (ingredientExistsError) {
            throw ingredientExistsError;
        };

        const { data: unitsQuery, error: unitsQueryError } = await supabaseClient
            .from('units')
            .select('id')
            .ilike('unit_name', units)
            .single();

        if (unitsQueryError) {
            throw unitsQueryError;
        };

        const unit_id = unitsQuery.id;



        // if entry exists, then we need to update it, reuse unit info from table
        if (ingredientExists.length > 0) {
            const { error: selectedIngredientError } = await supabaseClient
                .from('user_ingredients')
                .update({
                    p_quantity: quantity,
                    p_unit_id: unit_id
                })
                .eq('aliased_ingredient_name', ingredientName)
                .eq('user_id', userID);

            if (selectedIngredientError) {
                throw selectedIngredientError;
            };
        }
        // if entry doesn't exist, we need to make it with all relevant info, so get unit_id from units table
        else {

            const { error: selectedIngredientError } = await supabaseClient
                .from('user_ingredients')
                .insert([{
                    user_id: userID,
                    aliased_ingredient_name: ingredientName,
                    p_quantity: quantity,
                    p_unit_id: unit_id
                }]);

            if (selectedIngredientError) {
                throw selectedIngredientError;
            };
        }

        return { success: true };

    }
    catch (error) {
        console.error(error);
        return { 'error': error.message };
    }   

    // code for flask backend operation
    // const response = await fetch(`${API_BASE}/modify-ingredient-entry`, {method: 'POST', headers: { 'Content-Type': 'application/json' }, body: JSON.stringify({ ingredientName, quantity, units })});
    // return await response.json();
};

export default modifyIngredientEntry;